Rewrite lints page

GuillaumeGomez opened the pull request:
This PR has multiple goals:

* Make lints page to work without needing a web server by removing the json file.
* Prepare the field to also make the page work with JS (not done in this PR but should be straightforward).
* Remove angular dependency.

r? @Alexendoo 

changelog: make lint page work without web server

xFrednet commented on util/gh-pages/index_template.html:
```diff
@@ -0,0 +1,262 @@
+<!DOCTYPE html>
+<!--
+Welcome to a Clippy's lint list, at least the source code of it. If you are
+interested in contributing to this website checkout `util/gh-pages/index.html`
```
This comment should be updated to point to the template file

GuillaumeGomez replied:
Good catch, thanks!

Alexendoo commented on .github/deploy.sh:
```diff
@@ -9,7 +9,6 @@ echo "Making the docs for master"
 mkdir out/master/
 cp util/gh-pages/index.html out/master
 cp util/gh-pages/script.js out/master
-cp util/gh-pages/lints.json out/master
```
With https://github.com/rust-lang/rust-clippy/blob/cbf019b413417918ef1a89c9cef7d64c0c65d8e0/.github/workflows/deploy.yml#L48

Would this create an issue with deploying the 1.81 docs? cc @flip1995

GuillaumeGomez replied:
What issue do you expect?

Alexendoo replied:
If it uses the `index.html` of the time but the latest `deploy.sh` it would be missing the required `lints.json`

GuillaumeGomez replied:
Oh I see. An in-between situation.

flip1995 replied:
Yeah, we probably have to push old versions of the `html` files to the gh-pages` branch manually for the 2 releases following this. 

Thanks for the ping! This should block improvements here though. I figured something similar out before and am optimistic to be able to do that again :)

GuillaumeGomez replied:
Just to confirm: did you forget "not" in "This should block improvements"? :fearful:

Alexendoo commented on util/gh-pages/index_template.html:
```diff
@@ -0,0 +1,262 @@
+                                <li role="separator" class="divider"></li>
+                            </ul>
+                        </div>
+                        <div class="btn-group", id="lint-applicabilities" tabindex="-1">
```
Shouldn't be a comma here

GuillaumeGomez replied:
Arf, good catch!

Alexendoo commented on util/gh-pages/index_template.html:
```diff
@@ -0,0 +1,262 @@
+                            {% if let Some(id_location) = lint.id_location %}
+                                <div class="lint-additional-info-item">
+                                    <a href="https://github.com/rust-lang/rust-clippy/blob/master/clippy_lints/{(id_location)}">View Source</a>
+                            {% endif %}
+                            </div>
```
![image](https://github.com/user-attachments/assets/ee0ecf1b-eaf3-4e98-8e49-85e869992344)


```suggestion
                                </div>
                            {% endif %}
```

GuillaumeGomez commented:
Fixed both DOM issues.

Alexendoo commented:
I've found that it's slower to load than the current site, not so much on my desktop but it's noticeable on my laptop/phone. There may be some value in keeping the `lints.json` method around so we can lazily create the lint bodies as requested

Relatedly having to rerun `collect-metadata` after any change to the template isn't great

GuillaumeGomez commented:
It's slower but since I'm aiming at having the website work even with JS disabled, we can't do much about it unfortunately. I have some tricks up my sleeve to reduce the page size by a lot though. I wanted to send follow-ups since this PR is already quite big.

> Relatedly having to rerun collect-metadata after any change to the template isn't great

It's not really a big impact though. I don't expect many people will have this problem.

GuillaumeGomez commented:
It's the usual: do you prefer users accessibility or users comfort? And in this case, since I plan to make more changes, it might not even be one or the other once I'm done since like I said, I have a lot of others changes lined up. Let me describe them so maybe it'll help you in your decision:

 1. Transform all lints into `<details>` so the JS to collapse/expand can be removed.
 2. Enable auto-whitespace stripping in `rinja` so that all rinja tags (like `{% %}`, `{# #}`...) will strip all whitespace surrounding them by default (more information [here](https://rinja.readthedocs.io/en/stable/configuration.html#whitespace-control)).
 3. Generate the menus and anything that require JS to work in JS directly like we do in rustdoc to reduce even further the page size by default.

And final argument: this website won't even need a server anymore once this PR is merged to work locally since the JSON file won't exist anymore. Meaning we can even distribute the website very easily and open it locally with `cargo clippy doc` or something like that.

Alexendoo commented:
What accessibility are we referring to here? Assistive technologies have no issue with elements created via JS

That aside, it was a response to us not being able to do much about it, if that's not the case then it's completely fine to ditch the JSON file

I've had a closer look and it's not solely down to the increase in transfer size/HTML parsing. With a couple changes it should be on par with `master`

1. Make theme selection happen earlier (preferably inlined in `<head>`), currently the unthemed page becomes visible while the page is loading which makes the slowdown more noticeable
![image](https://github.com/user-attachments/assets/c64f643c-676f-4c19-83ea-00eb02c526ff)

2. Make syntax highlighting lazy, a significant amount of the time is spent here upfront when it could happen only when the section is expanded

Alexendoo commented:
> 1. Transform all lints into `<details>` so the JS to collapse/expand can be removed.

Personally I find the ctrl + F behaviour of `<details>` to be frustrating, auto expanding sections is not something I want it to do

> 2. Enable auto-whitespace stripping in `rinja` so that all rinja tags (like `{% %}`, `{# #}`...) will strip all whitespace surrounding them by default (more information [here](https://rinja.readthedocs.io/en/stable/configuration.html#whitespace-control)).

Sounds good

> 3. Generate the menus and anything that require JS to work in JS directly like we do in rustdoc to reduce even further the page size by default.

Seems like it would be a neutral change size wise if it's added to the JS, the menus are pretty small anyway

GuillaumeGomez commented:
> What accessibility are we referring to here? Assistive technologies have no issue with elements created via JS

Mostly smaller devices like e-reader will have issues with JS. But in any case, it's bad practice to have the rendering done in two passes (one loading JS which in turn generates HTML) as it prevents to have information on the page directly.

> I've had a closer look and it's not solely down to the increase in transfer size/HTML parsing. With a couple changes it should be on par with `master`

Yes, but that's something I wanted to do in another PR originally. Do you want me to do it in this PR directly? Also please note that when hosted on a web server (so not locally), all queries are compressed so the download time shouldn't be impacted much.

> Make theme selection happen earlier (preferably inlined in <head>), currently the unthemed page becomes visible while the page is loading which makes the slowdown more noticeable

Agreed, going to change it then (likely moving the theme-handling JS into a different file, loaded a the beginning).

> Make syntax highlighting lazy, a significant amount of the time is spent here upfront when it could happen only when the section is expanded

This is a very good idea!

> Personally I find the ctrl + F behaviour of <details> to be frustrating, auto expanding sections is not something I want it to do

At least in firefox it doesn't expand `<details>` when using ctrl+f. I'm surprised any web browser would do it.

> Seems like it would be a neutral change size wise if it's added to the JS, the menus are pretty small anyway

True. But I'll hide them if JS is disabled so at this point, why not just not generate them if the JS disabled. Less CSS required then. ;)

Alexendoo commented:
> Yes, but that's something I wanted to do in another PR originally. Do you want me to do it in this PR directly?

Yes please, I think we should hit parity before landing it

> when hosted on a web server (so not locally), all queries are compressed so the download time shouldn't be impacted much.

I did my testing on a server with it hosted alongside the `master` version, overall it transferred an extra ~30kb which yeah, is fine

> At least in firefox it doesn't expand `<details>` when using ctrl+f. I'm surprised any web browser would do it.

Unfortunately chrome does

GuillaumeGomez commented:
Ok, going to do it another way (like I did in [my blog](https://blog.guillaume-gomez.fr) for the "support me" button). But that's a task for another day anyway. Applying your suggestions in the meantime.

GuillaumeGomez commented:
Moved the theme handling code into its own file, removed some inlined CSS and greatly reduced the generated HTML page size:

```
before: 2.477.532 bytes
after:  1.777.286 bytes
```

GuillaumeGomez commented:
@Alexendoo Is there anything else to be done here?

Alexendoo commented on util/gh-pages/script.js:
```diff
@@ -1,629 +1,578 @@
+document.addEventListener("keypress", handleShortcut);
+document.addEventListener("keydown", handleShortcut);
+
+generateSettings();
+generateSearch();
```
Let's put these in the HTML from the start, it currently causes the page to move around a bit after a small delay

GuillaumeGomez replied:
This script is downloaded in a non-async method before the DOM is rendered, so there should not be any layout re-rendering (also I can't reproduce the bug locally, even with throttling).

GuillaumeGomez replied:
Good catch! The current lint page (before this PR) is displaying everything once it's loaded, so a different kind of "layout break". Seems like it's part of what I want to do as a follow-up: render buttons and any JS-bound element with JS instead of keeping it into HTML. Do you want me to do it in this PR directly or as a follow-up like I originally planned?

GuillaumeGomez commented on Cargo.toml:
```diff
@@ -39,6 +39,8 @@ toml = "0.7.3"
 filetime = "0.2.9"
 itertools = "0.12"
+pulldown-cmark = "0.11"
+rinja = { version = "0.3", default-features = false, features = ["config"] }
```
You might have heard of `askama`. I contributed to it a lot but some issues happened there (nothing bad, just different opinion on some specific things) and another `askama` maintainer and I decided to fork it.

You can see more information about it [here](https://blog.guillaume-gomez.fr/articles/2024-07-31+docs.rs+switching+jinja+template+framework+from+tera+to+rinja).

The biggest advantage of `rinja` is that if the template code is wrong, it will prevent compilation (which is the big reason why we moved to this framework for docs.rs).

GuillaumeGomez commented on util/gh-pages/index_template.html:
```diff
@@ -0,0 +1,232 @@
+<!DOCTYPE html>
+{#
+Welcome to a Clippy's lint list, at least the source code of it. If you are
+interested in contributing to this website checkout `util/gh-pages/index_template.html`
+inside the rust-clippy repository.
+
+Otherwise, have a great day =^.^=
+#}
```
Makes sense, changing it back.

GuillaumeGomez commented on util/gh-pages/style.css:
```diff
@@ -272,8 +272,6 @@ L4.75,12h2.5l0.5393066-2.1572876  c0.2276001-0.1062012,0.4459839-0.2269287,0.649
   height: 18px;
   display: block;
   filter: invert(0.7);
-  padding-left: 4px;
-  padding-top: 3px;
```
Strangely enough, if I put it back I get:

![image](https://github.com/user-attachments/assets/05cdb39c-8ffd-4975-a735-b8f6853258a9)

Alexendoo replied:
It is more noticeable in chrome (the UA stylesheet for button padding is different)
![image](https://github.com/user-attachments/assets/c242a83f-1c78-40bb-876a-dc68cd74bcef)

But the image is slightly below centre in firefox also, before/after: 
![image](https://github.com/user-attachments/assets/b55b935d-1429-4035-96e1-ad022ccc0bb7)
![image](https://github.com/user-attachments/assets/c0084095-64e9-4f21-a554-f688c2de60ea)

GuillaumeGomez replied:
Hum. Let me check if I can unify this across browsers.

GuillaumeGomez replied:
Instead of having something set by the default, I now force its position by using absolute positioning. I checked on firefox and chrome and both look the same now.

GuillaumeGomez commented:
Applied the suggestion for the HTML comment, for the rest, I can't reproduce the bugs locally...

GuillaumeGomez commented:
Fixed the settings button icon position.

bors commented:
:umbrella: The latest upstream changes (presumably #13384) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

GuillaumeGomez commented:
Fixed merge conflict.

bors commented:
:umbrella: The latest upstream changes (presumably #13440) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

GuillaumeGomez commented:
Fixed (new) merged conflict.

GuillaumeGomez commented:
@Alexendoo I moved the generation of the buttons before the DOM is loaded so it doesn't have the layout rework anymore.

Alexendoo commented:
(sorry didn't see [this reply](https://github.com/rust-lang/rust-clippy/pull/13269#discussion_r1766941134))

It still happens because the script appears after the elements in the DOM, it blocks the appearance of the lint list of lints but not the buttons

My preferred solution is to just put it in the HTML directly, you can think of it as manual SSR

GuillaumeGomez commented:
I did that: now the `script` is loaded before then I call the JS directly after the DOM for the settings menus is loaded. I tried with 3G throttling and there was no layout change.

Alexendoo commented:
I can still observe it on the latest commit. To solve these kind of layout changes you either hide the content until it's ready to be displayed (the old way) or have it be ready in the initial HTML

GuillaumeGomez commented:
Hiding them until they're ready sounds like a better idea. I replaced the two last commits with this change.

GuillaumeGomez commented:
It is in the original HTML. I kept the original behaviour where some filters have the number generated after for some reason. Anyway, I reverted to the previous version where the HTML is there and where the label is added with JS (like the current lints page).

GuillaumeGomez commented:
Got an idea on how to prevent this: I simply default numbers in the badges. They get updated when the JS loads but at least the layout doesn't change. Does it look good to you @Alexendoo ?

Alexendoo commented on util/gh-pages/index_template.html:
```diff
@@ -0,0 +1,232 @@
+                        </div> {# #}
+                        <div class="btn-group" id="lint-groups" tabindex="-1"> {# #}
+                            <button type="button" class="btn btn-default dropdown-toggle"> {# #}
+                                Lint groups <span class="badge">0</span> <span class="caret"></span> {# #}
```
```suggestion
                                Lint groups <span class="badge">9</span> <span class="caret"></span> {# #}
```

Alexendoo commented on util/gh-pages/index_template.html:
```diff
@@ -0,0 +1,232 @@
+                        <div class="btn-group", id="lint-applicabilities" tabindex="-1"> {# #}
+                            <button type="button" class="btn btn-default dropdown-toggle"> {# #}
+                                Applicability {#+ #}
+                                <span class="badge">0</span> {#+ #}
```
```suggestion
                                <span class="badge">4</span> {#+ #}
```

Alexendoo commented on util/gh-pages/index_template.html:
```diff
@@ -0,0 +1,232 @@
+                            </div>
+                            {# Clippy version #}
+                            <div class="lint-additional-info-item"> {# #}
+                                <span>{% if lint.group == "deprecated" %}Deprecated{% else %} Added{% endif %} in: </span> {# #}
```
Needs a <kbd>+</kbd> somewhere in here

![image](https://github.com/user-attachments/assets/9b6c3437-339a-4f2d-8d4c-bd424f855b02)

GuillaumeGomez replied:
Ah yes indeed!

Alexendoo commented on util/gh-pages/script.js:
```diff
@@ -1,629 +1,578 @@
+    // Expand the lint.
+    const lintElem = document.getElementById(lintId);
+    const isCollapsed = lintElem.classList.remove("collapsed");
+    lintElem.querySelector(".label-doc-folding").innerText = "-";
```
Should be `−` rather than `-`

GuillaumeGomez replied:
:+1:

Alexendoo commented on util/gh-pages/script.js:
```diff
@@ -1,629 +1,578 @@
+function expandLint(lintId) {
+    const lintElem = document.getElementById(lintId);
+    const isCollapsed = lintElem.classList.toggle("collapsed");
+    lintElem.querySelector(".label-doc-folding").innerText = isCollapsed ? "+" : "-";
```
Same here, but is this duplicated with the above?

GuillaumeGomez replied:
It is! Less code, yeay! \o/

Alexendoo commented on util/gh-pages/script.js:
```diff
@@ -1,629 +1,578 @@
-                };
-            };
+function highlightIfNeeded(elem) {
+    onEachLazy(elem.querySelectorAll("pre > code.language-rust:not(.highlighted)"), el => {
+        hljs.highlightElement(el)
+        el.classList.add("highlighted");
+    });
```
For the current styles I believe we'd want `.highlightElement(el.parentElement)`

![image](https://github.com/user-attachments/assets/0ead6884-b4fb-4e84-8d40-994626df99b0)
![image](https://github.com/user-attachments/assets/49d1b801-6688-428c-9d23-55222564f4f4)

GuillaumeGomez replied:
Wow good catch. I think once this PR is merged, I'll ask if the team would be interested into having a GUI testsuite for this page because you caught a lot of things and I'd prefer for this to not happen again in the future if we change things again.

GuillaumeGomez commented:
Applied your suggestions, thanks a lot for the thorough reviews. :)

Alexendoo commented:
Thanks, with a squash it looks good to go

GuillaumeGomez commented:
I squashed some changes and rewrote some parts of the git history so the changes can still be tracked independently. Or do you want just one commit?

Alexendoo commented:
That's 👍

@bors r+

bors commented:
:hourglass: Testing commit 603934336dea3c8385f0c7a03ebf2ecebfe913a2 with merge 6f1def79dd372e380357e1dfd2b8a54254ffe2d0...
<!-- homu: {"type":"BuildStarted","head_sha":"603934336dea3c8385f0c7a03ebf2ecebfe913a2","merge_sha":"6f1def79dd372e380357e1dfd2b8a54254ffe2d0"} -->

bors commented:
:sunny: Test successful - [checks-action_dev_test](https://github.com/rust-lang/rust-clippy/actions/runs/11294021791/job/31413572139), [checks-action_remark_test](https://github.com/rust-lang/rust-clippy/actions/runs/11294021788/job/31413579316), [checks-action_test](https://github.com/rust-lang/rust-clippy/actions/runs/11294021787/job/31414021908)
Approved by: Alexendoo
Pushing 6f1def79dd372e380357e1dfd2b8a54254ffe2d0 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"Alexendoo","base_ref":"master","builders":{"checks-action_test":"https://github.com/rust-lang/rust-clippy/actions/runs/11294021787/job/31414021908","checks-action_dev_test":"https://github.com/rust-lang/rust-clippy/actions/runs/11294021791/job/31413572139","checks-action_remark_test":"https://github.com/rust-lang/rust-clippy/actions/runs/11294021788/job/31413579316"},"merge_sha":"6f1def79dd372e380357e1dfd2b8a54254ffe2d0"} -->
